sympy/polys/matrices/domainmatrix.py

```diff
@@ -158,17 +162,22 @@ def __new__(cls, rows, shape, domain, *, fmt=None):
             else:
                 raise ValueError("fmt should be 'sparse' or 'dense'")
 
+        # Use python-flint for dense matrices if possible
+        if rep.fmt == 'dense' and DFM._supports_domain(domain):
+            rep = rep.to_dfm()
+
         return cls.from_rep(rep)
 
-    def __getnewargs__(self):
```

Comment:
This was deleted/renamed -- what is using this? I don't see this anywhere else in the diff and don't see use of `reduce(`, either.

Reply:
This was changed to `__reduce__`. The problem with `__getnewargs__` is that pickle still looks in `__dict__` and finds the `.rep` attribute which is a flint matrix. Flint matrices are not picklable so that fails.

Instead we need to implement `__reduce__` which gives full control over what objects are used for pickling. In this case we can just convert to list of lists format because flint's `fmpz` and `fmpq` are picklable.